internal/resolver/delegatingresolver/delegatingresolver.go

```diff
@@ -0,0 +1,304 @@
+			return nil, fmt.Errorf("delegating_resolver: unable to build the resolver for target %s: %v", target, err)
+		}
+	}
+
+	if r.proxyResolver, err = r.proxyURIResolver(opts); err != nil {
+		return nil, fmt.Errorf("delegating_resolver: failed to build resolver for proxy URL %q: %v", r.proxyURL, err)
+	}
+	return r, nil
+}
+
+// proxyURIResolver creates a resolver for resolving proxy URIs using the
+// "dns" scheme. It adjusts the proxyURL to conform to the "dns:///" format and
+// builds a resolver with a wrappingClientConn to capture resolved addresses.
+func (r *delegatingResolver) proxyURIResolver(opts resolver.BuildOptions) (resolver.Resolver, error) {
+	proxyBuilder := resolver.Get("dns")
+	if proxyBuilder == nil {
+		panic(fmt.Sprintln("delegating_resolver: resolver for proxy not found for scheme dns"))
+	}
+	r.proxyURL.Scheme = "dns"
+	r.proxyURL.Path = "/" + r.proxyURL.Host
+	r.proxyURL.Host = "" // Clear the Host field to conform to the "dns:///" format
+
+	proxyTarget := resolver.Target{URL: *r.proxyURL}
+	wcc := &wrappingClientConn{
+		parent:       r,
+		resolverType: proxyResolverType,
+	}
+	return proxyBuilder.Build(proxyTarget, wcc, opts)
+}
+
+func (r *delegatingResolver) ResolveNow(o resolver.ResolveNowOptions) {
+	if r.targetResolver != nil {
+		r.targetResolver.ResolveNow(o)
+	}
+	if r.proxyResolver != nil {
+		r.proxyResolver.ResolveNow(o)
+	}
+}
+
+func (r *delegatingResolver) Close() {
+	if r.targetResolver != nil {
+		r.targetResolver.Close()
+	}
+	if r.proxyResolver != nil {
+		r.proxyResolver.Close()
+	}
+}
+
+// generateCombinedAddressesLocked creates a list of combined addresses by
+// pairing each proxy address with every target address. For each pair, it
+// generates a new [resolver.Address] using the proxy address, and adding the
+// target address as the attribute along with user info.
+func (r *delegatingResolver) generateCombinedAddressesLocked() ([]resolver.Address, []resolver.Endpoint) {
+	var addresses []resolver.Address
+	for _, proxyAddr := range r.proxyAddrs {
+		for _, targetAddr := range r.targetAddrs {
+			newAddr := resolver.Address{Addr: proxyAddr.Addr}
+			newAddr = proxyattributes.Populate(newAddr, proxyattributes.Options{
+				User:        r.proxyURL.User,
+				ConnectAddr: targetAddr.Addr,
+			})
+			addresses = append(addresses, newAddr)
+		}
+	}
+
+	// Create a list of combined addresses by pairing each proxy endpoint
+	// with every target endpoint. For each pair, it constructs a new
+	// [resolver.Endpoint] using the first address from the proxy endpoint,
+	// as the DNS resolver is expected to return only one address per proxy
+	// endpoint. The target address and user information from the proxy URL
+	// are added as attributes to this address. The resulting list of addresses
+	// is then grouped into endpoints, covering all combinations of proxy and
+	// target endpoints.
+	var endpoints []resolver.Endpoint
```

Comment:
Based on my understanding, an endpoint is a list of addresses which all route to the same server. IMO, the logic for combining the proxy addresses with the target addresses should ensure all the addresses that route a particular server belong to the same endpoint.

This means that the endpoint grouping of the proxy resolver doesn't really matter while deciding the combined endpoints. So we can flatten all the proxy endpoints (or resolver.Addresses if endpoints are not set) into a single `[]resolver.Address{}` slice. For each address in a target endpoint, we create an address that uses each of the proxy addresses. The following pseudo code concisely shows this:
```
// Create a flat proxy address list from endpoints (or addresses if endpoints are not set).
proxy_addresses = []Address{}
if len(proxy_state.Endpoints) > 0:
    for ep in proxy_state.Endpoints:
        for addr in ep.Addresses:
            proxy_addresses.push(addr)
else:
    proxy_addresses = proxy_state.Addresses


for target_ep in target_resolver_state.Endpoints:
    combined_addresses = []Address{}
    for addr in target_ep.Addresses:
        for proxy_addr in proxy:
		    combined_addresses.push(addTarAttribute(proxy_addr, addr))
    target_ep.Addresses = combined_addresses
```

Note that this ensures that the number of endpoints produced is the same as the number of endpoints in the target resolver state.